libraries/common/ref.effekt

```diff
@@ -9,26 +9,69 @@ extern js """
   }
 """
 
+extern llvm """
+define void @c_ref_erase_field(ptr %0) {
+  %field_pointer = getelementptr inbounds ptr, ptr %0, i64 1
+  %tag = load i64, ptr %0, align 8
+  %field = load ptr, ptr %field_pointer, align 8
+  tail call void @erasePositive(i64 %tag, ptr %field)
+  ret void
+}
+"""
+
 /// Global, mutable references
 extern type Ref[T]
 
+/** We represent references like positive types in LLVM
+ *  The tag is 0 and the obj points to memory with the following layout:
+ *
+ *   +--[ Header ]--------------+------------+
+ *   | ReferenceCount  | Eraser | Field      |
+ *   +--------------------------+------------+
+ */
+
 /// Allocates a new reference, keeping its value _undefined_.
 /// Prefer using `ref` constructor instead to ensure that the value is defined.
 extern global def allocate[T](): Ref[T] =
   js "{ value: undefined }"
   chez "(box #f)"
   llvm """
-    %z = call %Pos @c_ref_fresh(%Pos zeroinitializer)
-    ret %Pos %z
+    ; sizeof Header + sizeof Pos = 32
+    %ref = tail call noalias ptr @malloc(i64 noundef 32)
```

Comment:
This must be `calloc`, because we might erase the reference before filling it.

EDIT: nevermind, I see you are storing `0` and `null` in the code below...